thetis/adr_model.py

```diff
@@ -0,0 +1,220 @@
+        pertaining to those species.
+    :type species: dict
+    """
+
+    # convert_xor causes "x^2" to be parsed as "x**2"
+    transformations = (
+        standard_transformations + (convert_xor,))
+
+    reserved_symbols = {
+        "pi", "e",
+        "x", "y", "z", "t",
+        "u", "u_x", "u_y", "u_z"}
+
+    def __init__(self, model_dict, lambdify_modules='numpy'):
+        """ADR_Model constructor.
+
+        Uses a user-defined dict to create a data structure containing
+        the names and symbols associated with all constants and species
+        in the ADR system, the values of all constants, and the
+        diffusivities and reaction terms of all species.
+
+        :arg model_dict: Dictionary defining species in an ADR system.
+            This must include the keys used to identify each species,
+            their full names, their diffusivities and their reaction
+            terms.
+            Any hard-coded numeric values should be defined under
+            model_dict['constants'], and the details of each species
+            should be defined under model_dict['species'].
+            A dictionary obtained from a correctly-formatted YAML file using
+            `yaml.safe_load` should be sufficient.
+
+            **Example**
+
+            .. code-block:: python
+
+                {
+                    'constants': {
+                        'D1': 8e-05,
+                        'D2': 4e-05,
+                        'k1': 0.024,
+                        'k2': 0.06
+                    },
+                    'species': {
+                        'a': {
+                            'diffusion': 'D1',
+                            'name': 'Tracer A',
+                            'reaction': '-a*b**2 + k1*(1-a)'
+                        },
+                        'b': {
+                            'diffusion': 'D2',
+                            'name': 'Tracer B',
+                            'reaction': 'a*b^2 - (k1+k2)*b'
+                        }
+                    }
+                }
+
+        :type model_dict: dict
+        :kwarg lambdify_modules: Optional argument to pass as the `modules`
+            parameter of `sympy.lambdify`.
+        :type lambdify_modules: str or dict, optional
+        """
+        self.constants = {
+            "pi": fd.pi,
+            "e": fd.e
+        }
+        self.species = {}
+        self.constant_keys = set(model_dict["constants"].keys())
+        self.species_keys = set(model_dict["species"].keys())
+        self._validate_keys()
+        for k, v in model_dict["constants"].items():
+            self.constants[k] = float(v)
+        for i, (k, v) in enumerate(model_dict["species"].items()):
+            # Parse reaction and diffusion terms
+            reaction_term = parse_expr(
+                str(v["reaction"]),
+                transformations=ADR_Model.transformations)
+            diffusion_term = parse_expr(
+                str(v["diffusion"]),
```

Comment:
Maybe better to check this is a string above, rather than converting it? Or is there a case where you would actually want to convert from a different type?